src/main/java/org/mariadb/jdbc/internal/util/RedirectionInfo.java

```diff
@@ -0,0 +1,118 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with this library; if not, write to Monty Program Ab info@montyprogram.com.
+ *
+ * This particular MariaDB Client for Java file is work
+ * derived from a Drizzle-JDBC. Drizzle-JDBC file which is covered by subject to
+ * the following copyright and notice provisions:
+ *
+ * Copyright (c) 2009-2011, Marcus Eriksson
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this list
+ * of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice, this
+ * list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * Neither the name of the driver nor the names of its contributors may not be
+ * used to endorse or promote products derived from this software without specific
+ * prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS  AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ *
+ */
+
+package org.mariadb.jdbc.internal.util;
+
+import org.mariadb.jdbc.HostAddress;
+
+public class RedirectionInfo {
+    private final HostAddress host;
+    private final String user;
+
+    public RedirectionInfo(HostAddress host, String user) {
+        this.host = host;
+        this.user = user;
+    }
+
+    public HostAddress getHost() {
+        return host;
+    }
+
+    public String getUser() {
+        return user;
+    }
+
+    /**
+    * Parse redirection info from a message return by server.
+    *
+    * @param msg  the string which may contain redirection information.
+    * @return RedirectionInfo host and user for redirection.
+    */
+    public static RedirectionInfo parseRedirectionInfo(String msg) {
+        /**
+         * Get redirected server information contained in OK packet.
+         * Redirection string somehow look like:
+         * Location: mysql://redirectedHostName:redirectedPort/user=redirectedUser
```

Comment:
A thought. Can this have `/?user=` instead of `/user=` ? That is
```
mysql://redirectedHostName:redirectedPort/?user=redirectedUser
```
that is, to make it a subset of a fairly common jdbc connection string syntax:
```
driver://host:port/database?key=val&key=val&...
```
this will allow to add new features easier without inventing new syntax every time. For example, you specify the redirectedUser here. It's possible that some day someone might want to have redirectedDatabase. Using jdbc url syntax — it's already there, no need to think how to integrate it in the url.

And it'll be easier to explain and document, because it's what people already know.

Reply:
I, too, thought this syntax was a little strange, especially when `&ttl=NNN` might be appended as another "parameter".

By using `/?user`, any client should be able to use a standard URL-parsing library to extract the parameters, instead of doing manual string parsing. It would also provide a standard answer (i.e., use URL-encoding) for what to do about "special" characters in the username, e.g., how to handle a username containing `&`.

Reply:
@vuvova @bgrainer would you please consider putting comments related to redirection string format to server side PR https://github.com/MariaDB/server/pull/1472 ? As the format is still subject to changes we would like to first change them on server side, and align them to client side after server side PR is completed and the format is settled.

Reply:
I will make the change on server side to use /?user= instead of /user=.

Reply:
Has updated the protocol to 
Location: mysql://[%s]:%u/?user=%s&ttl=%u\n